compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64StringUTF16CompressOp.java

```diff
@@ -0,0 +1,185 @@
+    @Temp({REG}) protected AllocatableValue vectorTemp3;
+
+    public AArch64StringUTF16CompressOp(LIRGeneratorTool tool, AllocatableValue src, AllocatableValue dst, AllocatableValue len, AllocatableValue result) {
+        super(TYPE);
+        this.src = src;
+        this.dst = dst;
+        this.len = len;
+        resultValue = result;
+        LIRKind archWordKind = LIRKind.value(tool.target().arch.getWordKind());
+        temp1 = tool.newVariable(archWordKind);
+        temp2 = tool.newVariable(archWordKind);
+        temp3 = tool.newVariable(archWordKind);
+        LIRKind vectorKind = LIRKind.value(tool.target().arch.getLargestStorableKind(SIMD));
+        vectorTemp1 = tool.newVariable(vectorKind);
+        vectorTemp2 = tool.newVariable(vectorKind);
+        vectorTemp3 = tool.newVariable(vectorKind);
+        chunkByteSize = 32;
+    }
+
+    @Override
+    public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
+        Label simdImpl = new Label();
+        Label failToCompress = new Label();
+        Label ret = new Label();
+
+        /* Initialise result for successful return */
+        masm.mov(64, asRegister(resultValue), asRegister(len));
+        masm.compare(64, asRegister(len), chunkByteSize);
+        masm.branchConditionally(ConditionFlag.GE, simdImpl);
+        emitScalar(masm, failToCompress);
+        masm.jmp(ret);
+
+        masm.bind(simdImpl);
+        emitSIMD(masm, failToCompress, ret);
+        masm.jmp(ret);
+
+        masm.bind(failToCompress);
+        masm.mov(64, asRegister(resultValue), zr);
+
+        masm.bind(ret);
+    }
+
+    private void emitScalar(AArch64MacroAssembler masm, Label failToCompress) {
+        Register srcAddress = asRegister(temp1);
+        Register destAddress = asRegister(temp2);
+        Register largestLatin1 = asRegister(temp3);
+
+        Label loop = new Label();
+
+        try (ScratchRegister scratchReg1 = masm.getScratchRegister(); ScratchRegister scratchReg2 = masm.getScratchRegister()) {
+            Register count = scratchReg1.getRegister();
+            Register currData = scratchReg2.getRegister();
+
+            masm.mov(64, srcAddress, asRegister(src));
+            masm.mov(64, destAddress, asRegister(dst));
+            masm.lsl(64, count, asRegister(len), 1);
+            masm.mov(largestLatin1, 0xFF);
+
+            masm.align(16);
+            masm.bind(loop);
+            /*
+             * If the current char is greater than 0xFF, stop compressing and return zero as result
+             */
+            masm.ldr(16, currData, AArch64Address.createImmediateAddress(16, AddressingMode.IMMEDIATE_POST_INDEXED, srcAddress, 2));
+            masm.cmp(64, currData, largestLatin1);
+            masm.branchConditionally(ConditionFlag.GT, failToCompress);
+
+            masm.str(8, currData, AArch64Address.createImmediateAddress(8, AddressingMode.IMMEDIATE_POST_INDEXED, destAddress, 1));
+            masm.subs(64, count, count, 2);
```

Comment:
shouldn't this be by 1? you already did a lsl on line 118

Reply:
Ah, sorry I somehow saw that as a shift right...

Now my question is what's the benefit of shift left and then subtracting by 2 instead of no shift and subtract by 1?

Reply:
no no, it slipped through refactoring. The previous version was using count as an offset instead of the post-index.